### PR TITLE
WIP: Kubernetes discovery service

### PR DESCRIPTION
**Pull request by hendrikebbers**
First try of kubernetes discovery service

---
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/866)

### CODE REVIEW

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            final CompletableFuture<URL> future = new CompletableFuture<>();
+            future.complete(cachedURL);
+            return future;
+        } else {
+            return httpClient.get(discoveryServiceEndpoint + "/services/" + name).withoutContent().readString().execute().handle((r, e) -> {
+                if (e != null) {
+                    throw new DolphinRuntimeException("Can not get address for name " + name, e);
+                }
+                try {
+                    URL url = new URL(r.getContent());
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                    cache.put(name, url);
+                    return url;
+                } catch (MalformedURLException e1) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            final CompletableFuture<URL> future = new CompletableFuture<>();
+            future.complete(cachedURL);
+            return future;
+        } else {
+            return httpClient.get(discoveryServiceEndpoint + "/services/" + namespace + "/" + name).withoutContent().readString().execute().handle((r, e) -> {
+                if (e != null) {
+                    throw new DolphinRuntimeException("Can not get address for namespace " + namespace + " and name " + name, e);
+                }
+                try {
+                    URL url = new URL(r.getContent());
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Override
+    public CompletableFuture<URL> getAddress(final String namespace, final String name) {
+        final URL cachedURL = cache.get(namespace + ":" + name);
+        if(cachedURL != null) {
+            final CompletableFuture<URL> future = new CompletableFuture<>();
+            future.complete(cachedURL);
+            return future;
+        } else {
+            return httpClient.get(discoveryServiceEndpoint + "/services/" + namespace + "/" + name).withoutContent().readString().execute().handle((r, e) -> {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java]: I would prefer to use a URL Builder to create the URL used with the httpClient. Concatenated strings are a bit of prone to errors.

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                if (e != null) {
+                    throw new DolphinRuntimeException("Can not get address for namespace " + namespace + " and name " + name, e);
+                }
+                try {
+                    URL url = new URL(r.getContent());
+                    cache.put(namespace + ":" + name, url);
+                    return url;
+                } catch (MalformedURLException e1) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/KubernetesURLStreamHandlerFactory.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        return Collections.singletonList(KUBERNETES_PROTOCOL);
+    }
+
+    @Override
+    public URLStreamHandler createURLStreamHandler(final String protocol) {
+        if(KUBERNETES_PROTOCOL.equals(protocol)) {
+            new URLStreamHandler(){
+
+                @Override
+                protected URLConnection openConnection(final URL u) throws IOException {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/KubernetesURLStreamHandlerFactory.java]: Would like to see `url` not `u` here :)

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/KubernetesURLStreamHandlerFactory.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                    Assert.requireNonNull(u, "u");
+                    final String serviceName = u.getHost();
+                    if(serviceName.contains(":")) {
+                        final String namespace = serviceName.split(":")[0];
+                        final String name = serviceName.split(":")[1];
+                        try {
+                            return discoveryService.getAddress(namespace, name).get().openConnection();
+                        } catch (Exception e) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/KubernetesURLStreamHandlerFactory.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/KubernetesURLStreamHandlerFactory.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+                            throw new IOException("Can not convert service URL", e);
+                        }
+                    } else {
+                        final String name = serviceName;
+                        try {
+                            return discoveryService.getAddress(name).get().openConnection();
+                        } catch (Exception e) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/KubernetesURLStreamHandlerFactory.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-kubernetes-discovery-server/src/main/java/com/canoo/dp/impl/platform/kubernetes/KubernetesClientConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    @Value("${dolphinPlatform.kubernetes.user}")
+    private String user;
+
+    @Value("${dolphinPlatform.kubernetes.password}")
+    private String password;
+
+    @Bean("kubernetesClient")
+    @ApplicationScope
+    public KubernetesClient createKubernetesClient() {
+        Config config = new ConfigBuilder()
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-kubernetes-discovery-server/src/main/java/com/canoo/dp/impl/platform/kubernetes/KubernetesClientConfiguration.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-kubernetes-discovery-server/src/main/java/com/canoo/dp/impl/platform/kubernetes/KubernetesDiscoveryServer.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.canoo.platform.server.spring.EnableDolphinPlatform;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+@EnableDolphinPlatform
+public class KubernetesDiscoveryServer {
+
+    public static void main(String[] args) {
```
Comment (giftkugel) [inline platform-extras/dolphin-platform-kubernetes-discovery-server/src/main/java/com/canoo/dp/impl/platform/kubernetes/KubernetesDiscoveryServer.java]: Add `final` keyword.

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java (hunk 1 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java]: done

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java (hunk 2 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java]: done

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java (hunk 3 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java]: done

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java (hunk 4 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/DiscoveryServiceImpl.java]: see https://github.com/canoo/dolphin-platform/issues/873

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/KubernetesURLStreamHandlerFactory.java (hunk 6 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/KubernetesURLStreamHandlerFactory.java]: done

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/KubernetesURLStreamHandlerFactory.java (hunk 7 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/KubernetesURLStreamHandlerFactory.java]: done

File: platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/KubernetesURLStreamHandlerFactory.java (hunk 8 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-kubernetes-discovery-client/src/main/java/com/canoo/dp/impl/KubernetesURLStreamHandlerFactory.java]: done

File: platform-extras/dolphin-platform-kubernetes-discovery-server/src/main/java/com/canoo/dp/impl/platform/kubernetes/KubernetesClientConfiguration.java (hunk 9 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-kubernetes-discovery-server/src/main/java/com/canoo/dp/impl/platform/kubernetes/KubernetesClientConfiguration.java]: done

File: platform-extras/dolphin-platform-kubernetes-discovery-server/src/main/java/com/canoo/dp/impl/platform/kubernetes/KubernetesDiscoveryServer.java (hunk 10 above)
Comment (hendrikebbers) [reply platform-extras/dolphin-platform-kubernetes-discovery-server/src/main/java/com/canoo/dp/impl/platform/kubernetes/KubernetesDiscoveryServer.java]: done

**Pull request closed without merging**